crates/core/tedge/src/cli/certificate/cli.rs

```diff
@@ -193,16 +196,187 @@ pub enum UploadCertCli {
     },
 }
 
-fn get_device_id_from_config(
+/// Returns the device ID from the config if no ID is provided by CLI
+/// If the provided device ID mismatches the one from the config, it returns an error
+fn get_device_id(
+    id: Option<String>,
     config: &TEdgeConfig,
+    config_location: &TEdgeConfigLocation,
     cloud: &Option<Cloud>,
-) -> Result<String, ReadError> {
-    let id = match cloud {
-        None => config.device.id(),
-        Some(Cloud::C8y(profile)) => config.c8y.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Azure(profile)) => config.az.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Aws(profile)) => config.aws.try_get(profile.as_deref())?.device.id(),
-    }?
-    .to_owned();
-    Ok(id)
+) -> Result<String, TEdgeError> {
+    let config_id = config.device_id(cloud.as_ref()).ok();
+    let writable_key = get_writable_key(cloud)?;
+    let dto = config_location
+        .load_dto_from_toml_and_env()
+        .context("failed to read tedge.toml file")?;
+
+    match id {
+        None => match config_id {
+            None => Err(anyhow!("No device ID is provided. Use `--device-id <name>` option to specify the device ID.").into()),
+            Some(config_id) => Ok(config_id.into()),
+        }
+        Some(input_id) if config_id.is_none() => Ok(input_id),
+        Some(input_id) if input_id == config_id.unwrap() => Ok(input_id),
+        Some(input_id) => match cloud.as_ref().map(<_>::into) {
```

Comment:
Matching on the `(id, config_id)` pair would make clearer the intent that is to compare the two ids checking for conflicts.

```rust
    match (id, config_id) {
        (None, None) => Err(anyhow!("No device ID is provided. Use `--device-id <name>` option to specify the device ID.").into()),
        (None, Some(config_id)) => Ok(config_id.into()),
        (Some(input_id), None) => Ok(input_id),
        (Some(input_id), Some(config_id)) if input_id == config_id => Ok(input_id),
        (Some(input_id), Some(config_id)) => {
            // ...
        }
    }
```

Reply:
Looks good. Since the condition is quite complex, more readability is definitely helpful.

Reply:
Simplified in 99620aa